uostr.3
.TH uostr 3
.SH NAME
uostr \- dynamically allocated strings
.SH SYNTAX
.B #include "uostr.h"
.RS
.nf
typedef struct uostr_t {
        char *data;
        size_t size;
        size_t len;
} \fBuostr_t\fP;
.fi
.RE
.sp
.BI "uostr_t * uostr_allocmore(uostr_t *u, size_t bytes);"
.br
.BI "uostr_t * uostr_xallocmore(uostr_t *u, size_t bytes);"
.br
.BI "void uostr_freedata(uostr_t *);"
.br
.BI "void uostr_forget(uostr_t *);"
.br
.BI "uostr_t *uostr_cut(uostr_t *,long new_len);"
.br
.BI "uostr_fcut(uostr_t *): /* i'm a macro */"
.br
.BI uostr_needmore(uostr_t *,size_t bytes)
.br
.BI uostr_xneedmore(uostr_t *,size_t bytes)
.br
.BI uostr_need(uostr_t *,size_t bytes)
.br
.BI uostr_xneed(uostr_t *,size_t bytes)
.PP
.BI "uostr_t *uostr_dup_cstr(uostr_t *,const char *);"
.br
.BI "uostr_t *uostr_xdup_cstr(uostr_t *,const char *);"
.br
.BI "uostr_t *uostr_add_cstr(uostr_t *,const char *);"
.br
.BI "uostr_t *uostr_xadd_cstr(uostr_t *,const char *);"
.PP
.BI "uostr_t *uostr_dup_cstrmulti(uostr_t *u,...);"
.br
.BI "uostr_t *uostr_xdup_cstrmulti(uostr_t *u,...);"
.br
.BI "uostr_t *uostr_add_cstrmulti(uostr_t *u,...);"
.br
.BI "uostr_t *uostr_xadd_cstrmulti(uostr_t *u,...);"
.PP
.BI "uostr_t *uostr_dup_char(uostr_t *,const char);"
.br
.BI "uostr_t *uostr_xdup_char(uostr_t *,const char);"
.br
.BI "uostr_t *uostr_add_char(uostr_t *,const char);"
.br
.BI "uostr_t *uostr_xadd_char(uostr_t *,const char);"
.PP
.BI "uostr_t *uostr_dup_uostr(uostr_t *,const uostr_t *);"
.br
.BI "uostr_t *uostr_xdup_uostr(uostr_t *,const uostr_t *);"
.br
.BI "uostr_t *uostr_add_uostr(uostr_t *,const uostr_t *);"
.br
.BI "uostr_t *uostr_xadd_uostr(uostr_t *,const uostr_t *);"
.PP
.BI "uostr_t *uostr_dup_mem(uostr_t *,const void *, size_t len);"
.br
.BI "uostr_t *uostr_xdup_mem(uostr_t *,const void *, size_t len);"
.br
.BI "uostr_t *uostr_add_mem(uostr_t *,const void *, size_t len);"
.br
.BI "uostr_t *uostr_xadd_mem(uostr_t *,const void *, size_t len);"
.br
.BI "void (*uostr_xallocfn)(const char *);" 

.SH DESCRIPTION
An
.B uostr_t
variable holds a string in dynamically allocated space.
The length of the string is unlimited (as long as
malloc() permits). ASCII-0 is allowed as part of 
the string and is not considered a terminating charakter
by any of this functions.
.br
The functions do not silently append a trailing ASCII-0. The see a C
string in the same way as the strlen() function does. You have to
explicitly append a trailing 0.

.br
The functions are reasonable fast and safe against
buffer overflows, you do not need to worry about
overflowing the string boundaries. 

The structure consists of the parts:
.I data
is the pointer to the data,
.I size
is the amount of memory allocated for the string (you should not care about it) and
.I len
is the length of the string (which may be of interest to you).

To create a string use something like 
.B uostr_t my_string={0}; .
You need to return the memory to the system at a later time, use
.B uostr_freedata(&my_string) 
for that.

The functions often allocate some bytes more than they need at the moment (they
try to be clever and allocate a useful amount of memory), so there is often
almost no performance penalty if you append something.
.br
If more space is needed the string may be moved in memory. Do not use pointers
to the 
.I uostr_t.data
until you have finished adding things to it.

Many functions come in two flavors: one being a normal function
which returns a
.I NULL
in the
.B out of memory
situations, and the other one do not return in that case. The 
naming convention is straightforward: if the name of the 
function or macro contains an
.B x
it will not return. Use that with care. Normally those functions
will terminate the program after they print 
.I out of memory in uostr_FUNCTION_NAME 
followed by a line feed. To change this behaviour use the 
.B uostr_xallocfn 
variable, see below.


.I string
is just a name, an 
.B uostr_t 
may actually contain whatever you want.

.SH USAGE

.I uostr_t * uostr_allocmore(uostr_t *u, size_t bytes);
allocates more memory to the string. It returns 
.B 0
in case it cannot get more memory, but does not change
anything. 
.I uostr_t * uostr_xallocmore(uostr_t *u, size_t bytes);
basically does the same thing, but does not return if it 
cannot allocate more memory.

.I uostr_needmore(uostr_t *,size_t bytes)
and 
.I uostr_xneedmore(uostr_t *,size_t bytes)
are macros which call the allocmore functions
if more memory needs to be allocated. They are
somewhat more efficient than calling the
functions "just in case".

.I uostr_need(uostr_t *,size_t bytes)
and
.I uostr_xneed(uostr_t *,size_t bytes)
are macros to make sure that 
.B bytes
bytes of memory are available for that
string (you may want to use them if you
know that the string will grow a lot).

.I void uostr_freedata(uostr_t *s);
frees the memory allocated to a string and 
sets s->data to NULL.

.I "void uostr_forget(uostr_t *);"
tells the allocation routines of the uostr library
to forget about that string. Usually they bark load
("reusing using" + abort()) if they suspect a memory
leak, which can lead to false alerts in cases like this:
.EX
.nf
          char *overcomplicated_strdup(const char *text) 
          {
           uostr_t s=UOSTR_INIT;
		   uostr_xadd_cstr(&s,text);
           return s.data;
          }
.fi
.EE
Usually the second call to the function would fail because
uostr_freedata() has not been called, but inserting an
.I "uostr_forget(&s)"
make the routines forget about that string (set the global
variable 
.B "int uostr_debug_free"
to 0 to disable this behaviour completely).

.I uostr_t *uostr_cut(uostr_t *,long new_len);
cuts the length of a string. If 
.I new_len
is a positive integer the string will be cut to
that amount of characters, otherwise it will be
cut by that many characters.
.br
This function is often used with 
.B uostr_t.len
values safed before, see the example below.

Note that the long should be a ssize_t, but that's
not portable, and other functions in this library
do not need a config.h and/or autoconf.

.I uostr_fcut(uostr_t *);
cuts the string to a length of exactly 0 characters.

The following functions allocate more memory if they need
it, and they all have a 
.B x 
variant:

.I uostr_t *uostr_dup_cstr(uostr_t *,const char *);
overwrites the string with the 
.B C 
language 0 terminated string, but will
.B not
duplicate the trailing ASCII-0.

.I uostr_t *uostr_add_cstr(uostr_t *,const char *);
appends the 0 terminated string to the uostr_t. It
will not duplicate the trailing ASCII-0.

.I uostr_t *uostr_dup_cstrmulti(uostr_t *u,...);
overwrites the string with a concatenation of the arguments.
You need to terminate the argument list with a 
.B NULL .
Use this for example for:
.br
.I uostr_dup_cstrmulti(&s,varname,"=",value,NULL);
.br
Use with care, it's easy to forget the trailing 
NULL pointer. Not that this function does not
terminate the string with an ASCII-0.

.I uostr_t *uostr_add_cstrmulti(uostr_t *u,...);
appends the string with the concatenation.

.I uostr_t *uostr_dup_char(uostr_t *,const char);
overwrites the string with a single character.

.I uostr_t *uostr_add_char(uostr_t *,const char);
appends a single character to the string.

.I uostr_t *uostr_dup_uostr(uostr_t *,const uostr_t *);
overwrites one uostr with the data of the other.

.I uostr_t *uostr_add_uostr(uostr_t *,const uostr_t *);
append one string to the other.

.I uostr_t *uostr_dup_mem(uostr_t *,const void *, size_t len);
overwrites the string with an arbitrary amount of 
characters from the memory. This is the basic function all
.I uostr_dup_something 
functions are based on.

.I uostr_t *uostr_add_mem(uostr_t *,const void *, size_t len);
appends an arbitrary amount of characters from the memory to the string. 
This is the basic function all
.I uostr_add_something 
functions are based on.

.I uostr_0(uostr_t *)
and
.I uostr_x0(uostr_t *)
are simple macros which append a ASCII-0 (C string functions 
recognize this as string terminator) to a string.

.I uostr_xallocfn
is a variable defined like this:

.B "extern void (*uostr_xallocfn)(const char *);"

This function will, if it contains a value different from the
NULL pointer (which behaves in all practical aspects very much
like the 0), call another function to terminate the program. It
will get one argument, the name of the function which failed
to allocate memory. Please not that your function must call
exit.

The default function will simple print 
.I "out of memory in uostr_FUNCTION_NAME"
followed by a line feed.

.SH EXAMPLE
Error handling omitted.

  uostr_t path={0};
  DIR *d;
  struct dirent *de;
  uostr_xdup_cstr(&path,"/var/log");
  uostr_x0(&path); /* terminate that \0 */
  d=opendir(path.data);
  uostr_cut(&path,-1); /* throw away that \0 */
  uostr_xadd_char(&path,'/');
  while ((de=readdir(d))!=NULL) {
    size_t pos=path.len;
    uostr_xadd_cstr(path,de->d_name);
    uostr_x0(&path); /* terminate that \0 */
    unlink(path.data); XXX do not run me!
    uostr_cut(&path,pos);
  }
  closedir(d);
  uostr_freedata(&path); 

.SH ERROR HANDLING
The 
.I x
functions will not return in case the run out of memory.

.I uostr_{add,dup}_cstrmulti
will return NULL if it cannot allocate more memory, but may already have
changed the string.

All other functions return NULL if they fail to allocate the memory they need and 
will not touch the string.

.SH "SEE ALSO"
.I bugtraq 
to see why overflow safe strings are useful.

.SH COPYRIGHT
Copyright 1998,1999 Uwe Ohse.
.br
This routines are published under the GNU
General Public License,
see
.IR COPYING 
for a full statement of rights and permissions.
.SH AUTHOR
Uwe Ohse


NEWS
0.2.4 - 1 November 1999

* some cleanup and minor bug fixes
* utftpd left SCCS lock files laying around in case of a failed put
  request.

0.2.3 - 5 September 1999

* the server now ignores garbage bytes at the end of a packet, unless
  there were tftp options.
* "must not write as root" check moved some lines up, to check against
  this before we create files (old version left 0 length files).

0.2.2 - 22 August 1999

* the server now uses adaptive timeout.
* the server now unlinks the received file if it was not received 
  completely.
* the server now includes a workaround for the 
  "Sorcerer's Apprentice Syndrome" (as it's called by RFC 1123)

0.2.1 - 2 May 1999

* rewrote configuration file handling. It's now much more 
  beautiful. 
  See sample.config for more information.
  The old stuff partially continues to work, except that
* the \ at the end of line (to split long lines into smaller pieces)
  will not work anymore. Sorry for that, but it's just too much
  pain to fix this.

0.2.0 - 2 May 1999
* understands a new experimental tftp option (revision).
* major internal interface cleanup.
* now reads version controlled files from a pipe. This means that
  it now can serve read request even when the same file is just uploaded.
* "lying around" copies of version controlled files will now be ignored.
* -l option (GNU inetutils) is now recognized (and ignored,
  utftpd logs by default).
* -n option (GNU inetutils) implemented. It's used to suppress NAK
  answers for "file not found". 
* Can now operate without a configuration file, in a mode largely 
  compatible with the traditional UNIX tftp server: Access control
  based on a list of directories given on the command line and the
  UNIX access bits (read/write for others). Therefore
* it may now be used as a drop-in replacement for GNU inetutils tftpd.

0.1.5 - 10 April 1999
* added RCS support. See the utftpd.conf manual page for a description 
  of the new configuration variables rcs-ci and rcs-co.
* improved SCCS support: unchecked files will not be checked in 
  (that's useful if your router saves his configuration regulary even 
  if it's unchanged). See the utftpd.conf manual page for a 
  description of the new configuration file variables sccs-clean and
  sccs-unget.
* improved checksuite to check SCCS/RCS handling.
* fixed some small bugs.
* improved compatibility (SCO 3.2v4.2, BSDI 3.1, Solaris 2.6 and 2.5.1,
  OpenBSD 2.4/alpha, HPUX 9.00/m68k).

0.1.4 - 5 April 1999
* configuration (cdb) file may now reside outside the chrooted area.
* understands RFC 2348 "blksize" option.
* understands RFC 2349 "timeout" option (but doesn't know about "tsize" 
  now).
* test suite added.
* documentation fixes.
* server now knows about wildcards.
* threw out some bugs and unused library files.
* there's now a client program (utftp).

0.1.3 - 28 March 1999
* uh, 0.1.2 didn't work because of those stupid last-moment-large-rewrite ... 
  Ok, that's why i tested it on my companies tftp server.
* smaller log things.

0.1.2 - 28 March 1999
* manual pages
* rpm fixed

0.1.1 - 21 March 1999
* put on cvs


uostr.c
/*
 * Copyright (C) 1998,1999 Uwe Ohse
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * 
 * As a special exception this source may be used as part of the
 * SRS project by CORE/Computer Service Langenbach
 * regardless of the copyright they choose.
 * 
 * Contact: uwe@ohse.de
 */
#include <stdlib.h>
#include <errno.h>
#include <unistd.h> /* write */
#include "uostr.h"

/* core functions */

	/* assume max. overhead of malloc() */
#define OVERHEAD (sizeof(void *)+sizeof(size_t))

#define DEBUG_FREE

#ifdef DEBUG_FREE
int uostr_debug_free=1;
struct deb {
	uostr_t *adr;
	struct deb *next;
};
static struct deb *anker;
#endif

void (*uostr_xallocfn) P__((const char *))=0;
void 
uostr_xallocerr(const char *fn)
{
	if (uostr_xallocfn) uostr_xallocfn(fn);
	else {
		const char *p=fn; while(*p!=0) p++;
		write(2,"out of memory in ",17);
		write(2,fn,p-fn);
		write(2,"\n",1);
	}
	exit(1);
}

uostr_t *
uostr_allocmore(uostr_t *u, size_t bytes)
{
	size_t n;
	size_t t;
	if (!u->data) n=bytes;
	else n=bytes+u->size;
	for (t=32;;t*=2) {
		if (t-OVERHEAD>=n)
			break;
	}
	t-=OVERHEAD;
	if (!u->data) {
		u->len=0;
#ifdef DEBUG_FREE
		if (uostr_debug_free) {
			struct deb *neu,*x;	
			for (x=anker;x;x=x->next) { if (x->adr==u) break; }
			if (x) { write(2,"reusing uostr\n",14); abort(); }
			else {
				neu=(struct deb *) malloc(sizeof(struct deb));
				if (!neu) {
					uostr_debug_free=0;
					while (anker) { struct deb *next; next=anker->next; free(anker); anker=next; }
				} else { neu->adr=u; neu->next=anker; anker=neu; }
			}
		}
#endif
		u->data=(char *) malloc(t);
		if (!u->data) { errno=ENOMEM; return 0;}
	}
	else {
		char *tmp=(char *) realloc(u->data,t);
		if (!tmp) return 0;
		u->data=tmp;
	}
	u->size=t;
	return u;
}
uostr_t *
uostr_xallocmore(uostr_t *u, size_t bytes)
{ uostr_t *r=uostr_allocmore(u,bytes); if (r) return r; uostr_xallocerr("uostr_xallocmore"); return(0); }

void
uostr_freedata(uostr_t *u)
{
	if (u->data) {
#ifdef DEBUG_FREE
		if (uostr_debug_free) {
			struct deb *x,*l;
			for (x=anker,l=0;x;) {
				if (x->adr==u) {
					struct deb *n=x->next;
					free(x);
					if (l) l->next=n;
					else anker=n;
					x=n;
					continue;
				}
				l=x;
				x=x->next;
			}
		}
#endif
		free(u->data);
		u->data=0;
	}
}
void
uostr_forget(uostr_t *u)
{
#ifdef DEBUG_FREE
	if (uostr_debug_free) {
		struct deb *x,*l;
		for (x=anker,l=0;x;) {
			if (x->adr==u) {
				struct deb *n=x->next;
				free(x);
				if (l) l->next=n;
				else anker=n;
				x=n;
				continue;
			}
			l=x;
			x=x->next;
		}
	}
#endif
}
